design/vela-core/environment.md

```diff
@@ -0,0 +1,144 @@
+# Managing Environments As Shared Bases For Application Deployment
+
+## Background
+
+A team of application development usually needs to prepare some shared environments for developers to deploy applications.
+For example, most teams will have two environments, a *dev* environment for testing their applications in live instances,
+and a *prod* environment for deploying applications to serve live requests. The environment is a logical group bundling common resources that multiple application deployments might depend on.
+
+Here are some examples of what users want to define in an environment:
+
+- K8s Cluster. *dev* environment might have smaller clusters while *prod* might use large and complex clusters.
+- System components and definitions. *dev* environment might install different Operators & CRDs from *prod* environment, e.g. *dev* might choose local logging while *prod* might use ELK logging.
+- Network boundary. *dev* and *prod* environments would have different VPC setups, gateway endpoints, and firewall settings.
+- Secret and config. *dev* might use locally-run MySQL, while *prod* might use managed RDS. The DB connection credentials would be saved differently.
+- Policies. This is more important in *prod* environment that it would usually have more restrictions on application deployments, like security scanning, config checks, SLO checks, etc.
+
+All of above examples and more are bundled as a shared base for applications to run on.
+This is usualy done by an admin for the entire application team.
+By adding Environment API, Vela aligns with user workflow better,
+provides consistent view and pluggable options for end users,
+and empowering platform builders to add modular environment capabilities. In the following, we will propose the design of API and give an implementation plan.
+
+
+## Proposal
+
+### 1. Environment CRD
+
+We propose to add an Environment CRD to describe the shared resources for applications to base on.
+
+```yaml
+kind: Environment
+metadata:
+  name: prod
+spec:
+  # A list of pluggable EnvConfig resources.
+  # The resources will be parsed and applied when Environment CR is appled.
+  configs:
+    - type: cluster
+      properties:
+        name: prod-cluster
+        selector:
+          tier: prod
```

Comment:
In `Env`, What we talked about most is `cluster`. I am curious about what's the definition of `cluster`. Could you please set an example for type `cluster`:)

Reply:
+1

And what if it's not a K8s cluster? It's a env pointing to a cloud?